### PR TITLE
Implement User Search backend

### PR DESCRIPTION
**Pull request by mstruve**
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
Hooks up the backend for searching users. This will be followed up by a frontend search PR.

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/projects/6#card-34388442

## Added tests?
- [x] yes

## Added to documentation?
- [x] no documentation needed

alt_text

### CODE REVIEW

File: app/services/search/user.rb [hunk 1]
```diff
@@ -3,10 +3,49 @@ class User < Base
...
+        hits = results.dig("hits", "hits").map do |user_doc|
+          prepare_doc(user_doc.dig("_source"))
+        end
+        paginate_hits(hits, params)
+      end
+
       private
 
+      def prepare_doc(hit)
+        {
```
Comment (mstruve) [inline app/services/search/user.rb]: Right now this is the format the frontend expects so I am going to stick with it. Once everything is moved over we can work to make this cleaner and more practical

**Comment by mstruve** [comment]
CodeClimate has some valid refactoring recommendations. I think we should tackle those but not in this PR. Made a ticket for it: https://github.com/thepracticaldev/dev.to/projects/6#card-34490522

File: app/views/service_worker/index.js.erb [hunk 2]
```diff
@@ -118,6 +118,8 @@
         // Don't run on search endpoints
         !url.href.includes('/search/tags') &&
         !url.href.includes('/search/chat_channels') &&
+        !url.href.includes('/search/classified_listings') &&
```
Comment (atsmith813) [inline app/views/service_worker/index.js.erb]: Thanks 😅

File: app/services/search/user.rb [hunk 3]
```diff
@@ -3,10 +3,49 @@ class User < Base
...
+        params[:per_page] ||= DEFAULT_PER_PAGE
+
+        # pages start at 0
+        params[:size] = params[:per_page].to_i * (params[:page].to_i + 1)
+      end
+
+      def paginate_hits(hits, params)
+        start = params[:per_page] * params[:page]
+        hits[start, params[:per_page]] || []
+      end
```
Comment (atsmith813) [inline app/services/search/user.rb]: Not for this PR, but making a note we should consider moving this into the `Base` class.

File: spec/services/search/query_builders/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
+require "rails_helper"
```
Comment (atsmith813) [inline spec/services/search/query_builders/user_spec.rb]: Love how much smaller these spec files can be now with the refactor 🙏

**Pull request merged**